controllers/costmanagement_controller.go

```diff
@@ -262,6 +273,105 @@ func GetAuthSecret(r *CostManagementReconciler, costInput *CostManagementInput,
+	h.Set("Content-Type", "application/vnd.redhat.hccm.tar+tgz")
+	fw, err := mw.CreatePart(h)
+	f, err := os.Open(filename)
+	if err != nil {
+		log.Info("error opening file", err)
+	}
+	defer f.Close()
+	_, err = io.Copy(fw, f)
+	if err != nil {
+		log.Error(err, "The following error occurred")
+	}
+	mw.Close()
+	return buf, mw
+}
+
+func Upload(r *CostManagementReconciler, costInput *CostManagementInput, method string, path string, body *bytes.Buffer, mw *multipart.Writer) (string, string, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("costmanagement", "Upload")
+	req, err := http.NewRequest(method, path, body)
+	if err != nil {
+		log.Error(err, "Could not create request")
+		return "", "", err
+	}
+	// Create the header
+	if req.Header == nil {
+		req.Header = make(http.Header)
+	}
+	req = req.WithContext(ctx)
+	req.Header.Set("Content-Type", mw.FormDataContentType())
+	if costInput.Authentication == "basic" {
+		log.Info("Uploading using basic authentication!")
+		req.SetBasicAuth(costInput.BasicAuthUser, costInput.BasicAuthPassword)
+	} else {
+		log.Info("Uploading using token authentication")
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", costInput.BearerTokenString))
+		req.Header.Set("User-Agent", fmt.Sprintf("cost-mgmt-operator/%s cluster/%s", costInput.OperatorCommit, costInput.ClusterID))
+	}
+	// Log the headers - probably remove this later
+	log.Info("Request Headers:")
+	for key, val := range req.Header {
+		fmt.Println(key, val)
+	}
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Error(err, "Could not send request")
+		return "", "", err
+	}
+	defer resp.Body.Close()
+
+	fmt.Println("HTTP Response Status:", resp.StatusCode, http.StatusText(resp.StatusCode))
+	uploadStatus := fmt.Sprintf("%d ", resp.StatusCode) + string(http.StatusText(resp.StatusCode))
+	uploadTime := time.Now()
+
+	// Add error handling and logging here
+	requestID := resp.Header.Get("x-rh-insights-request-id")
+	if resp.StatusCode == http.StatusUnauthorized {
+		log.Info(fmt.Sprintf("gateway server %s returned 401, x-rh-insights-request-id=%s", resp.Request.URL, requestID))
+	}
+	if resp.StatusCode == http.StatusForbidden {
+		log.Info(fmt.Sprintf("gateway server %s returned 403, x-rh-insights-request-id=%s", resp.Request.URL, requestID))
+	}
+	if resp.StatusCode == http.StatusBadRequest {
+		body, _ := ioutil.ReadAll(resp.Body)
+		if len(body) > 1024 {
+			body = body[:1024]
+		}
+		log.Info(fmt.Sprintf("gateway server bad request: %s (request=%s): %s", resp.Request.URL, requestID, string(body)))
+	}
+
+	if resp.StatusCode >= 300 || resp.StatusCode < 200 {
+		body, _ := ioutil.ReadAll(resp.Body)
+		if len(body) > 1024 {
+			body = body[:1024]
+		}
+		log.Info(fmt.Sprintf("gateway server reported unexpected error code: %d (request=%s): %s", resp.StatusCode, requestID, string(body)))
+	}
+
+	if resp.StatusCode == http.StatusAccepted {
```

Comment:
So recently the ingress service changed its status response code and broke the insights operator. I think it might be "safer" consider >= 200 and < 300 as a success.

Reply:
That's a good call - I will change it!